src/Catalyst.Core.Modules.Keystore/KeyStoreService.cs

```diff
@@ -0,0 +1,839 @@
+                {
+                    Password = _dek
+                };
+                var pw = new PemWriter(sw);
+                pw.WriteObject(pkcs8);
+                await pw.Writer.FlushAsync();
+                pem = sw.ToString();
+            }
+
+            // Store the key in the repository.
+            var key = new EncryptedKey
+            {
+                Id = keyId.ToBase58(),
+                Name = name,
+                Pem = pem
+            };
+            await Store.PutAsync(name, key, cancel).ConfigureAwait(false);
+            log.DebugFormat("Added key '{0}' with ID {1}", name, keyId);
+
+            return new KeyInfo
+            {
+                Id = key.Id, Name = key.Name
+            };
+        }
+
+        /// <summary>
+        ///     Create a key ID for the key.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        /// <remarks>
+        ///     The key id is the SHA-256 multihash of its public key. The public key is
+        ///     a protobuf encoding containing a type and
+        ///     the DER encoding of the PKCS SubjectPublicKeyInfo.
+        /// </remarks>
+        private MultiHash CreateKeyId(AsymmetricKeyParameter key)
+        {
+            var spki = SubjectPublicKeyInfoFactory
+               .CreateSubjectPublicKeyInfo(key)
+               .GetDerEncoded();
+
+            // Add protobuf cruft.
+            var publicKey = new PublicKey
+            {
+                Data = spki
+            };
+            switch (key)
+            {
+                case RsaKeyParameters _:
+                    publicKey.Type = KeyType.RSA;
+                    break;
+                case ECPublicKeyParameters _:
+                    publicKey.Type = KeyType.Secp256k1;
+                    break;
+                case Ed25519PublicKeyParameters _:
+                    publicKey.Type = KeyType.Ed25519;
+                    break;
+                default:
+                    throw new NotSupportedException($"The key type {key.GetType().Name} is not supported.");
+            }
+
+            using (var ms = new MemoryStream())
+            {
+                Serializer.Serialize(ms, publicKey);
+
+                // If the length of the serialized bytes <= 42, then we compute the "identity" multihash of 
+                // the serialized bytes. The idea here is that if the serialized byte array 
+                // is short enough, we can fit it in a multihash verbatim without having to 
+                // condense it using a hash function.
+                var alg = ms.Length <= 48 ? "identity" : "sha2-256";
+
+                ms.Position = 0;
+                return MultiHash.ComputeHash(ms, alg);
+            }
+        }
+
+        private AsymmetricCipherKeyPair GetKeyPairFromPrivateKey(AsymmetricKeyParameter privateKey)
+        {
+            AsymmetricCipherKeyPair keyPair = null;
+            switch (privateKey)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add a 'default' clause to this 'switch' statement.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)